File: services/geo/geo/core_snapshot.py
"""Source selection and snapshot candidates. Storage activation is not done here."""
from .core_contract import bundled_schema, fail, index_records, parse_core, ref_key, require_revision, version_key
from .core_identity import validate_core_identity_graph
from .core_sources import is_precise_core_locator, unique, validate_core_source_catalog
from .core_frames import validate_core_frame_catalog
from .core_geometry import validate_core_geometry_catalog
from .core_signature import canonical_core_text, core_input_digest
import math


def _pin(value):
    return f"{ref_key(value['ref'])}@{value['revision']}"


def _version(value):
    return {"ref": value["ref"], "revision": value["revision"]}


def _records(values):
    return sorted(values, key=_pin)


def _bag(values):
    return sorted(values, key=lambda value: canonical_core_text(value).encode("utf-16-be"))


def _same_entity(a, b):
    return ref_key(a["entity"]) == ref_key(b["entity"])


def _prepare(value):
    canonical_core_text(value)
    data = parse_core("snapshot-input", value)
    validate_core_identity_graph(data["identity"])
    validate_core_source_catalog(data["sources"], data["identity"])
    validate_core_frame_catalog(data["frames"])
    validate_core_geometry_catalog(data["geometry"], data["identity"], data["sources"], data["frames"])
    worlds, observations, resolutions = [index_records(data[key], key) for key in ("worlds", "observations", "resolutions")]
    entities = index_records(data["identity"]["entities"], "entity")
    representations = index_records(data["geometry"]["representations"], "representation")
    quantities = index_records(data["geometry"]["reportedQuantities"], "reported quantity")
    parts, sources, assets, datasets = [index_records(data["sources"][key], key) for key in ("parts", "sources", "assets", "datasets")]
    frames = index_records(data["frames"]["frames"], "frame")
    index_records(data["compositions"], "composition")
    world = require_revision(worlds, data["context"]["world"], "snapshot world")
    policy = bundled_schema("snapshot-input")[0]["x-ulpin-snapshot-policy"]
    rank = policy["accessRank"]
    ceiling = rank[data["context"]["scope"]["ceiling"]]
    for item in data["sources"]["datasets"] + data["sources"]["assets"] + data["sources"]["sources"] + data["sources"]["parts"] + data["observations"]:
        if rank[item["access"]] > ceiling:
            fail("ACCESS_SCOPE", "Input is outside the declared authorization scope")
    asset_ranks = {}

    def asset_rank(link):
        root = require_revision(assets, link, "source asset")
        stack = [(root, False)]
        while stack:
            asset, leave = stack.pop()
            key = _pin(asset)
            if key in asset_ranks:
                continue
            if leave:
                asset_ranks[key] = max([rank[asset["access"]]] + [asset_ranks[_pin(parent)] for parent in asset["parentAssets"]])
                continue
            stack.append((asset, True))
            stack.extend((require_revision(assets, parent, "asset ancestry"), False) for parent in asset["parentAssets"] if _pin(parent) not in asset_ranks)
        return asset_ranks[_pin(root)]

    for observation in data["observations"]:
        if ref_key(observation["entity"]) not in entities:
            fail("MISSING_TARGET", "Observation entity is absent")
        observation_world = require_revision(worlds, observation["world"], "observation world")
        start, end = observation["validity"]["fromMs"], observation["validity"]["toMs"]
        if start is not None and end is not None and start >= end:
            fail("TIME_RANGE", "Validity must have an increasing half-open interval")
        unique([_pin(link) for link in observation["sourceParts"]], "observation source parts")
        required_rank = 0
        if observation_world["state"] == "observed" and observation["method"] == "synthetic":
            fail("WORLD_CLASSIFICATION", "Synthetic observations cannot be labelled observed")
        for link in observation["sourceParts"]:
            part = require_revision(parts, link, "observation source part")
            source = require_revision(sources, part["source"], "observation source")
            dataset = require_revision(datasets, source["dataset"], "source dataset") if source["dataset"] else None
            required_rank = max([required_rank, rank[part["access"]], rank[source["access"]], rank[dataset["access"]] if dataset else 0] + [asset_rank(asset) for asset in source["assets"]])
            if observation_world["state"] == "observed" and (source["method"] == "synthetic" or dataset and dataset["classification"] in ("planned", "hypothetical", "synthetic")):
                fail("WORLD_CLASSIFICATION", "Source classification contradicts an observed-world claim")
            if observation["payload"]["kind"] in ("geometry", "vertical_interval") and not any(is_precise_core_locator(locator) for locator in part["locators"]):
                fail("LOCATOR_NOT_QUALIFIED", "Geometry/vertical contribution requires an exact source part")
        payload = observation["payload"]
        part_refs = {_pin(link) for link in observation["sourceParts"]}
        if payload["kind"] == "geometry":
            rep = require_revision(representations, payload["representation"], "observed representation")
            if not _same_entity(rep, observation) or rep["role"] != observation["role"]:
                fail("OBSERVATION_ROLE", "Observation must match representation entity and role")
            if any(_pin(link) not in part_refs for link in rep["sourceParts"]):
                fail("OBSERVATION_PROVENANCE", "Observation omitted a geometry contribution source")
            if rep["geometry"]["profile"] == "asset":
                required_rank = max(required_rank, asset_rank(rep["geometry"]["asset"]))
        elif payload["kind"] == "vertical_interval":
            if observation["role"] != "vertical_interval":
                fail("OBSERVATION_ROLE", "Vertical payload has a different semantic role")
            frame = require_revision(frames, payload["frame"], "vertical observation frame")
            interval = payload["interval"]
            if interval is not None:
                if interval["upperMetres"] < interval["lowerMetres"] or not math.isfinite(interval["upperMetres"] - interval["lowerMetres"]):
                    fail("GEOMETRY_INTERVAL", "Vertical bounds are reversed or overflowed")
                if frame["kind"] == "geocentric" or frame["vertical"]["kind"] not in ("benchmark", "datum"):
                    fail("VERTICAL_UNRESOLVED", "Vertical interval requires a named reference")
                if _pin(interval["reference"]) != _pin(frame["vertical"]["reference"]):
                    fail("VERTICAL_MISMATCH", "Vertical observation and frame have different zeros")
        elif payload["kind"] == "reported_quantity":
            quantity = require_revision(quantities, payload["quantity"], "observed quantity")
            if not _same_entity(quantity, observation) or quantity["definition"] != observation["role"]:
                fail("OBSERVATION_ROLE", "Reported observation mismatches its definition/entity")
            if _pin(quantity["sourcePart"]) not in part_refs:
                fail("OBSERVATION_PROVENANCE", "Reported quantity source is missing")
        if rank[observation["access"]] < required_rank:
            fail("ACCESS_DOWNGRADE", "Observation cannot lower its contributing sources' access")
    resolution_keys = []
    for resolution in data["resolutions"]:
        if ref_key(resolution["entity"]) not in entities:
            fail("MISSING_TARGET", "Resolution entity is absent")
        require_revision(worlds, resolution["world"], "resolution world")
        unique([_pin(link) for link in resolution["candidates"]], "resolution candidates")
        for link in resolution["candidates"]:
            observation = require_revision(observations, link, "resolution candidate")
            if not _same_entity(observation, resolution) or _pin(observation["world"]) != _pin(resolution["world"]) or observation["role"] != resolution["role"]:
                fail("RESOLUTION_SCOPE", "Candidates differ in entity, world or semantic role")
        if resolution["selected"] and _pin(resolution["selected"]) not in {_pin(link) for link in resolution["candidates"]}:
            fail("RESOLUTION_SELECTION", "Selected observation is not an exact candidate")
        resolution_keys.append(canonical_core_text([resolution["entity"], resolution["world"], resolution["role"], resolution.get("purpose", "analysis")]))
    unique(resolution_keys, "entity/world/role resolution")
    generated = set()
    for composition in data["compositions"]:
        if ref_key(composition["entity"]) not in entities:
            fail("MISSING_TARGET", "Composition entity is absent")
        require_revision(worlds, composition["world"], "composition world")
        links = [composition["geometry"]] if composition["kind"] == "passthrough" else [composition["footprint"], composition["vertical"]]
        for link in links:
            resolution = require_revision(resolutions, link, "composition resolution")
            if not _same_entity(resolution, composition) or _pin(resolution["world"]) != _pin(composition["world"]):
                fail("COMPOSITION_SCOPE", "Composition resolution has another world or entity")
        if composition["kind"] == "prism":
            footprint = require_revision(resolutions, composition["footprint"], "footprint resolution")
            vertical = require_revision(resolutions, composition["vertical"], "vertical resolution")
            if footprint["role"] not in policy["footprintRoles"] or vertical["role"] != "vertical_interval":
                fail("COMPOSITION_ROLE", "Prism requires qualified footprint and vertical schedule")
            key = ref_key(composition["output"]["ref"])
            if key in representations or key in generated:
                fail("OUTPUT_ID_COLLISION", "Generated geometry cannot replace an existing identity")
            generated.add(key)
    return data, world, observations, resolutions, representations, quantities, policy


def validate_core_snapshot_input(value):
    return _prepare(value)[0]


def _ordered_input(data):
    entities = []
    for entity in _records(data["identity"]["entities"]):
        lifecycle = entity["lifecycle"]
        entities.append({**entity, "identifiers": _bag(entity["identifiers"]), "memberships": _bag(entity["memberships"]),
                         "lifecycle": lifecycle if lifecycle["state"] == "active" else {**lifecycle, "replacedBy": _bag(lifecycle["replacedBy"])}})
    sources = data["sources"]
    ordered_sources = {**sources, "datasets": _records(sources["datasets"]),
                       "assets": [{**a, "parentAssets": _bag(a["parentAssets"])} for a in _records(sources["assets"])],
                       "sources": [{**s, "assets": _bag(s["assets"]), "workflows": _bag(s["workflows"])} for s in _records(sources["sources"])],
                       "parts": _records(sources["parts"]), "links": _records(sources["links"])}
    if "linkHistory" in sources:
        ordered_sources["linkHistory"] = _records(sources["linkHistory"])
    reported = [{**q, "amount": {**q["amount"], "candidates": _bag(q["amount"]["candidates"])} if q["amount"]["state"] == "conflicting" else q["amount"]} for q in _records(data["geometry"]["reportedQuantities"])]
    return {**data, "worlds": _records(data["worlds"]),
            "identity": {"entities": entities, "relations": sorted(data["identity"]["relations"], key=lambda r: r["id"])},
            "sources": ordered_sources, "frames": {"frames": _records(data["frames"]["frames"]), "operations": _records(data["frames"]["operations"])},
            "geometry": {"representations": [{**r, "sourceParts": _bag(r["sourceParts"])} for r in _records(data["geometry"]["representations"])], "reportedQuantities": reported},
            "observations": [{**o, "sourceParts": _bag(o["sourceParts"])} for o in _records(data["observations"])],
            "resolutions": [{**r, "candidates": _bag(r["candidates"])} for r in _records(data["resolutions"])], "compositions": _records(data["compositions"])}


def build_core_snapshot(value):
    data, world, observations, resolutions, representations, quantities, policy = _prepare(value)
    selected, selected_quantities, results = {}, {}, []

    def choose(link):
        resolution = require_revision(resolutions, link, "selected resolution")
        if resolution["selected"] is None:
            return None, "UNRESOLVED_SELECTION", "unknown"
        observation = require_revision(observations, resolution["selected"], "selected observation")
        start, end, time = observation["validity"]["fromMs"], observation["validity"]["toMs"], data["context"]["asOfMs"]
        coverage = "bounded" if time is not None and start is not None and end is not None else "unknown"
        if time is not None and (start is not None and time < start or end is not None and time >= end):
            return None, "OUTSIDE_VALIDITY", coverage
        if observation["payload"]["kind"] == "unavailable":
            return None, observation["payload"]["reasonCode"], coverage
        return observation, None, coverage

    for composition in _records(data["compositions"]):
        if _pin(composition["world"]) != _pin(data["context"]["world"]):
            continue
        base = {"composition": _version(composition), "entity": composition["entity"]}
        def fail_result(reason, coverage="unknown"):
            results.append({**base, "status": "unavailable", "representation": None, "reasonCode": reason, "temporalCoverage": coverage})
        first, reason, coverage = choose(composition["geometry"] if composition["kind"] == "passthrough" else composition["footprint"])
        if first is None:
            fail_result(reason, coverage)
            continue
        if first["payload"]["kind"] != "geometry":
            fail("COMPOSITION_PAYLOAD", "Geometry selection has no geometry payload")
        rep = require_revision(representations, first["payload"]["representation"], "selected geometry")
        output = rep
        if composition["kind"] == "prism":
            second, reason, second_coverage = choose(composition["vertical"])
            if second is None:
                fail_result(reason, second_coverage)
                continue
            vertical = second["payload"]
            if vertical["kind"] != "vertical_interval":
                fail("COMPOSITION_PAYLOAD", "Vertical selection has no interval payload")
            geometry = rep["geometry"]
            shape = geometry["geometry"] if geometry["profile"] == "planar" else geometry["footprint"] if geometry["profile"] == "prism" else None
            if shape is None or shape["type"] not in ("Polygon", "MultiPolygon"):
                fail_result("POLYGON_REQUIRED")
                continue
            if rep["frame"] is None:
                fail_result("FRAME_UNRESOLVED")
                continue
            if _pin(rep["frame"]) != _pin(vertical["frame"]):
                fail_result("FRAME_MISMATCH")
                continue
            if vertical["interval"] is None:
                fail_result("UNKNOWN_VERTICAL_INTERVAL")
                continue
            parts = {_pin(link): link for link in first["sourceParts"] + second["sourceParts"]}
            output = {**composition["output"], "entity": composition["entity"], "frame": rep["frame"], "sourceParts": _records(list(parts.values())),
                      "geometry": {"profile": "prism", "footprint": shape, "interval": vertical["interval"]}}
            coverage = "bounded" if coverage == "bounded" and second_coverage == "bounded" else "unknown"
        selected[ref_key(output["ref"])] = output
        results.append({**base, "status": "available", "representation": _version(output), "reasonCode": None, "temporalCoverage": coverage})
    for resolution in data["resolutions"]:
        if _pin(resolution["world"]) != _pin(data["context"]["world"]):
            continue
        choice, _, _ = choose(_version(resolution))
        if choice is None or choice["payload"]["kind"] != "reported_quantity":
            continue
        quantity = require_revision(quantities, choice["payload"]["quantity"], "selected quantity")
        selected_quantities[ref_key(quantity["ref"])] = quantity
        if quantity["amount"]["state"] == "conflicting":
            for link in quantity["amount"]["candidates"]:
                candidate = require_revision(quantities, link, "selected quantity candidate")
                selected_quantities[ref_key(candidate["ref"])] = candidate
    geometry = validate_core_geometry_catalog({"representations": _records(list(selected.values())), "reportedQuantities": _records(list(selected_quantities.values()))}, data["identity"], data["sources"], data["frames"])
    used = {ref_key(r["entity"]) for r in geometry["representations"] + geometry["reportedQuantities"]}
    entities = _records([e for e in data["identity"]["entities"] if ref_key(e["ref"]) in used])
    input_digest = core_input_digest(_ordered_input(data))
    geometry_digest = core_input_digest({"profile": policy["compilerProfile"], "world": world,
                                         "frames": {"frames": _records(data["frames"]["frames"]), "operations": _records(data["frames"]["operations"])},
                                         "entities": [{"ref": e["ref"], "kind": e["kind"]} for e in entities],
                                         "representations": [{key: value for key, value in r.items() if key != "sourceParts"} for r in geometry["representations"]]})
    manifest = parse_core("snapshot-manifest", {"schemaVersion": "ulpin-core-snapshot/1", "state": "candidate", "context": data["context"],
                                               "signatureVersion": policy["canonicalEncoding"], "inputDigest": input_digest, "geometryDigest": geometry_digest,
                                               "representations": [_version(r) for r in geometry["representations"]], "entities": [_version(e) for e in entities],
                                               "retainedObservations": [_version(o) for o in _records(data["observations"]) if _pin(o["world"]) == _pin(data["context"]["world"])]})
    return {"manifest": manifest, "geometry": geometry, "results": results}


def validate_core_publication_candidate(snapshot_value, publication_value):
    snapshot = parse_core("snapshot-manifest", snapshot_value)
    publication = parse_core("publication-candidate", publication_value)
    if publication["snapshotDigest"] != snapshot["inputDigest"] or publication["geometryDigest"] != snapshot["geometryDigest"]:
        fail("PUBLICATION_SNAPSHOT", "Publication belongs to another immutable snapshot")
    if canonical_core_text(publication["scope"]) != canonical_core_text(snapshot["context"]["scope"]):
        fail("ACCESS_SCOPE", "Publication authorization scope differs")
    unique([a["id"] for a in publication["assets"]], "publication assets")
    assets = {a["id"] for a in publication["assets"]}
    reps = {_pin(r) for r in snapshot["representations"]}
    unique([canonical_core_text([b["assetId"], b["featureId"]]) for b in publication["bindings"]], "asset feature identities")
    for binding in publication["bindings"]:
        if binding["assetId"] not in assets:
            fail("MISSING_REFERENCE", "Publication feature refers to an absent derived asset")
        if _pin(binding["representation"]) not in reps:
            fail("STALE_REFERENCE", "Feature is not an exact selected representation revision")
    return publication

File: tests/fixtures/core-snapshot.ts
import {geometryFixture} from "./core-geometry";
import {ref} from "./core-identity";
import {buildCoreSnapshot} from "../../packages/contracts/src/spatial/core/snapshot";

export const snapshotRef=(namespace:string,id:string,revision=1)=>({ref:{namespace,id},revision});
export function snapshotFixture() {
  const g=geometryFixture(),world=snapshotRef("world","two-block-fixture");
  const base=g.geometry.representations[0];
  const footprint={...base,ref:{namespace:"representation",id:"footprint"},role:"ground_footprint",geometry:{profile:"planar",geometry:base.geometry.footprint}};
  const unit={...base,ref:{namespace:"representation",id:"unit-plan"},entity:ref("U1"),role:"unit_boundary",sourceParts:[snapshotRef("source_part","unit-part")],geometry:{profile:"planar",geometry:{type:"Polygon",coordinates:[[[0,0],[10,0],[10,4],[0,4],[0,0]]]}}};
  const original=g.sources.assets[0],source=g.sources.sources[0],part=g.sources.parts[0];
  const sources={...g.sources,
    assets:[original,{...original,ref:ref("height-asset","asset"),sha256:"b".repeat(64),mediaType:"text/csv",storage:{state:"registered",blobRef:ref("height-bytes","source_blob")}},
      {...original,ref:ref("unit-asset","asset"),sha256:"c".repeat(64),storage:{state:"registered",blobRef:ref("unit-bytes","source_blob")}}],
    sources:[source,{...source,ref:ref("height-source","source_revision"),family:ref("height-family","source_family"),familyOrdinal:1,profile:"level-schedule",assets:[snapshotRef("asset","height-asset")]},
      {...source,ref:ref("unit-source","source_revision"),family:ref("unit-family","source_family"),familyOrdinal:1,profile:"unit-plan",assets:[snapshotRef("asset","unit-asset")]}],
    parts:[part,{...part,ref:ref("height-part","source_part"),source:snapshotRef("source_revision","height-source"),asset:snapshotRef("asset","height-asset"),locators:[{kind:"rows",range:{start:1,end:2}}]},
      {...part,ref:ref("unit-part","source_part"),source:snapshotRef("source_revision","unit-source"),asset:snapshotRef("asset","unit-asset"),locators:[{kind:"feature",featureId:"U1"}]}],
  };
  const observe=(id:string,entity:string,role:string,sourceParts:unknown[],payload:unknown)=>({ref:ref(id,"observation"),revision:1,entity:ref(entity),world,role,sourceParts,payload,method:"synthetic",access:"operator",validity:{fromMs:0,toMs:2000}});
  const observations=[
    observe("footprint","B1","ground_footprint",footprint.sourceParts,{kind:"geometry",representation:snapshotRef("representation","footprint")}),
    observe("height-six","B1","vertical_interval",[snapshotRef("source_part","height-part")],{kind:"vertical_interval",frame:base.frame,interval:base.geometry.interval}),
    observe("height-nine","B1","vertical_interval",[snapshotRef("source_part","height-part")],{kind:"vertical_interval",frame:base.frame,interval:{...base.geometry.interval,upperMetres:9}}),
    observe("unit","U1","unit_boundary",unit.sourceParts,{kind:"geometry",representation:snapshotRef("representation","unit-plan")}),
  ];
  const resolve=(id:string,entity:string,role:string,candidates:string[],selected:string)=>({ref:ref(id,"resolution"),revision:1,entity:ref(entity),world,role,candidates:candidates.map(id=>snapshotRef("observation",id)),selected:snapshotRef("observation",selected),reason:"Explicit authored fixture selection, not a latest-file rule"});
  return {
    schemaVersion:"ulpin-spatial/2",context:{world,asOfMs:1000,scope:{id:"local-operator",revision:1,ceiling:"operator"}},
    worlds:[{...world,label:"Synthetic two-block composition",state:"synthetic"}],identity:g.identity,frames:g.frames,sources,
    geometry:{representations:[footprint,unit],reportedQuantities:[]},observations,
    resolutions:[resolve("footprint","B1","ground_footprint",["footprint"],"footprint"),resolve("height","B1","vertical_interval",["height-six","height-nine"],"height-six"),resolve("unit","U1","unit_boundary",["unit"],"unit")],
    compositions:[
      {ref:ref("keep-footprint","composition"),revision:1,world,entity:ref("B1"),kind:"passthrough",geometry:snapshotRef("resolution","footprint")},
      {ref:ref("build-exterior","composition"),revision:1,world,entity:ref("B1"),kind:"prism",footprint:snapshotRef("resolution","footprint"),vertical:snapshotRef("resolution","height"),output:{ref:ref("exterior","representation"),revision:1,role:"exterior"}},
      {ref:ref("keep-unit","composition"),revision:1,world,entity:ref("U1"),kind:"passthrough",geometry:snapshotRef("resolution","unit")},
    ],
  };
}

export interface SnapshotCase {id:string;input:unknown;valid:boolean;code?:string;representationIds?:string[];reportedIds?:string[];volume?:number;unavailable?:string;temporalCoverage?:string}
export function snapshotCases():SnapshotCase[] {
  const rows:SnapshotCase[]=[];
  const good=(id:string,mutate:(x:any)=>void=()=>{},expected:Partial<SnapshotCase>={})=>{const input=JSON.parse(JSON.stringify(snapshotFixture()));mutate(input);rows.push({id,input,valid:true,representationIds:["exterior","footprint","unit-plan"],volume:480,...expected});};
  const bad=(id:string,code:string,mutate:(x:any)=>void)=>{const input=JSON.parse(JSON.stringify(snapshotFixture()));mutate(input);rows.push({id,input,valid:false,code});};
  good("footprint-schedule-unit-source-composition");
  good("explicit-other-height-not-newest-source",x=>x.resolutions[1].selected=snapshotRef("observation","height-nine"),{volume:720});
  good("failed-height-retains-footprint-and-unit",x=>x.observations[1].payload={kind:"unavailable",reasonCode:"HEIGHT_PARSE_FAILED",reason:"Original retained; failed schedule contribution"},{representationIds:["footprint","unit-plan"],volume:undefined,unavailable:"HEIGHT_PARSE_FAILED"});
  good("missing-height-keeps-valid-2d",x=>x.observations[1].payload.interval=null,{representationIds:["footprint","unit-plan"],volume:undefined,unavailable:"UNKNOWN_VERTICAL_INTERVAL"});
  good("unresolved-selection-not-guessed",x=>x.resolutions[1].selected=null,{representationIds:["footprint","unit-plan"],volume:undefined,unavailable:"UNRESOLVED_SELECTION"});
  good("half-open-validity-rejects-end-bound",x=>x.observations[1].validity.toMs=1000,{representationIds:["footprint","unit-plan"],volume:undefined,unavailable:"OUTSIDE_VALIDITY"});
  good("unknown-validity-is-explicit",x=>x.observations.forEach((o:any)=>o.validity={fromMs:null,toMs:null}),{temporalCoverage:"unknown"});
  good("bounded-validity-is-explicit",()=>{},{temporalCoverage:"bounded"});
  good("zero-source-documents-still-compose",x=>{x.sources={datasets:[],assets:[],sources:[],parts:[],links:[]};x.geometry.representations.forEach((r:any)=>r.sourceParts=[]);x.observations.forEach((o:any)=>o.sourceParts=[]);});
  good("same-label-different-frame-does-not-compose",x=>{x.frames.frames.push({...x.frames.frames[0],ref:ref("other","frame")});x.observations[1].payload.frame=snapshotRef("frame","other");},{representationIds:["footprint","unit-plan"],volume:undefined,unavailable:"FRAME_MISMATCH"});
  good("unplaced-footprint-not-placed-by-height",x=>x.geometry.representations[0].frame=null,{representationIds:["footprint","unit-plan"],volume:undefined,unavailable:"FRAME_UNRESOLVED"});
  const quantities=(x:any)=>{
    const sourcePart=snapshotRef("source_part","p1");
    x.geometry.reportedQuantities=[{ref:ref("gross","reported_quantity"),revision:1,entity:ref("B1"),definition:"gross_floor_area",unit:"m2",amount:{state:"known",value:100},sourcePart},
      {ref:ref("net","reported_quantity"),revision:1,entity:ref("B1"),definition:"net_floor_area",unit:"m2",amount:{state:"known",value:70},sourcePart}];
    for(const q of x.geometry.reportedQuantities){
      const observation={...x.observations[0],ref:ref(q.ref.id,"observation"),role:q.definition,payload:{kind:"reported_quantity",quantity:{ref:q.ref,revision:1}}};
      x.observations.push(observation);
      x.resolutions.push({...x.resolutions[0],ref:ref(q.ref.id,"resolution"),role:q.definition,candidates:[snapshotRef("observation",q.ref.id)],selected:snapshotRef("observation",q.ref.id)});
    }
  };
  good("reported-net-and-gross-remain-independent",quantities,{reportedIds:["gross","net"]});
  good("reported-conflict-preserves-required-known-candidates",x=>{
    quantities(x);const a=x.geometry.reportedQuantities[0];
    const b={...a,ref:ref("gross-other","reported_quantity"),amount:{state:"known",value:120}};
    const c={...a,ref:ref("gross-conflict","reported_quantity"),amount:{state:"conflicting",reason:"Source disagreement",candidates:[{ref:a.ref,revision:1},{ref:b.ref,revision:1}]}};
    x.geometry.reportedQuantities.push(b,c);x.observations[4].payload.quantity={ref:c.ref,revision:1};
  },{reportedIds:["gross-conflict","gross-other","gross","net"].sort()});
  good("other-world-does-not-leak-selected-features",x=>{
    const world=snapshotRef("world","empty-world");x.worlds.push({...world,label:"Separate empty scenario",state:"planned"});x.context.world=world;
  },{representationIds:[],volume:undefined});
  bad("stale-snapshot-world","STALE_REFERENCE",x=>x.context.world.revision=2);
  bad("observation-stale-world","STALE_REFERENCE",x=>x.observations[0].world={...x.observations[0].world,revision:2});
  bad("missing-observation-target","MISSING_TARGET",x=>x.observations[1].entity=ref("missing"));
  bad("stale-representation-observation","STALE_REFERENCE",x=>x.observations[0].payload.representation.revision=2);
  bad("missing-geometry-provenance","OBSERVATION_PROVENANCE",x=>x.observations[0].sourceParts=[]);
  bad("vertical-cannot-use-whole-file-locator","LOCATOR_NOT_QUALIFIED",x=>x.sources.parts[1].locators=[{kind:"whole_asset"}]);
  bad("wrong-observation-role","OBSERVATION_ROLE",x=>x.observations[0].role="roof_projection");
  bad("wrong-vertical-role","OBSERVATION_ROLE",x=>x.observations[1].role="ground_footprint");
  bad("bad-time-range","TIME_RANGE",x=>x.observations[1].validity={fromMs:10,toMs:10});
  bad("synthetic-not-an-observed-world","WORLD_CLASSIFICATION",x=>x.worlds[0].state="observed");
  bad("source-dataset-cannot-be-relabeled-observed","WORLD_CLASSIFICATION",x=>{x.worlds[0].state="observed";x.observations.forEach((o:any)=>o.method="source");x.sources.sources.forEach((s:any)=>s.method="source");});
  bad("public-scope-cannot-read-private-input","ACCESS_SCOPE",x=>x.context.scope.ceiling="public");
  bad("declared-observation-access-cannot-downgrade","ACCESS_DOWNGRADE",x=>x.observations[0].access="public");
  bad("restricted-asset-ancestor-cannot-be-hidden","ACCESS_DOWNGRADE",x=>{
    x.context.scope.ceiling="restricted";
    x.sources.assets.push({...x.sources.assets[0],ref:ref("private-parent","asset"),access:"restricted",storage:{state:"registered",blobRef:ref("private-original","source_blob")}});
    x.sources.assets[0].kind="derived";x.sources.assets[0].retention={...x.sources.assets[0].retention,policy:"managed_derivative"};x.sources.assets[0].parentAssets=[snapshotRef("asset","private-parent")];
  });
  bad("stale-selected-observation","RESOLUTION_SELECTION",x=>x.resolutions[1].selected.revision=2);
  bad("stale-candidate-observation","STALE_REFERENCE",x=>x.resolutions[1].candidates[0].revision=2);
  bad("wrong-entity-resolution","RESOLUTION_SCOPE",x=>x.resolutions[1].entity=ref("U1"));
  bad("duplicate-selection-policy","DUPLICATE_REFERENCE",x=>x.resolutions.push({...x.resolutions[0],ref:ref("other-policy","resolution")}));
  bad("selected-noncandidate","RESOLUTION_SELECTION",x=>x.resolutions[1].selected=snapshotRef("observation","unit"));
  bad("different-world-candidate","RESOLUTION_SCOPE",x=>{const world=snapshotRef("world","other-world");x.worlds.push({...world,label:"Other scenario",state:"synthetic"});x.observations[1].world=world;});
  bad("different-entity-composition","COMPOSITION_SCOPE",x=>x.compositions[1].entity=ref("U1"));
  bad("wrong-composition-source-role","COMPOSITION_ROLE",x=>x.compositions[1].footprint=snapshotRef("resolution","height"));
  bad("output-cannot-overwrite-original-representation","OUTPUT_ID_COLLISION",x=>x.compositions[1].output.ref.id="footprint");
  bad("two-generators-cannot-allocate-one-id","OUTPUT_ID_COLLISION",x=>x.compositions.push({...x.compositions[1],ref:ref("other-generator","composition")}));
  bad("unsupported-composition-cannot-smuggle-code","INVALID_CONTRACT",x=>x.compositions[1].script="return guessed_geometry()");
  bad("unpaired-unicode-is-not-a-portable-signature","SIGNATURE_UNICODE",x=>x.worlds[0].label="\ud800");
  return rows;
}

// Encodings are manually specified; digests were independently computed with hashlib.
export const signatureCases=[
  {id:"null",value:null,expectedEncoding:"z",expectedDigest:"594e519ae499312b29433b7dd8a97ff068defcba9755b6d5d00e84c524d67b06"},
  {id:"one",value:1,expectedEncoding:"n3ff0000000000000",expectedDigest:"73af61a19efb83c5922f1e2d07eceae40f55d2c60d0ffe1664f75e66f8f6c8da"},
  {id:"fraction",value:1.5,expectedEncoding:"n3ff8000000000000",expectedDigest:"673d4b6d67133a213da9eb3a2f2b42b0fa66ddd94c03e59787a1e75b60d26ccf"},
  {id:"record",value:{b:true,a:1},expectedEncoding:"o2:s1:an3ff0000000000000s1:bt",expectedDigest:"064e0488881c8b4ec8f5ec135a8e51f71061cef522f6221458529ac9dab9ffca"},
  {id:"utf8",value:"\u00e9",expectedEncoding:"s2:\u00e9",expectedDigest:"0566c2f70ee0c7b722fdbd0dae5e0e79e0e84ab169d7956b688da7823e471650"},
  {id:"array",value:[null,true,""],expectedEncoding:"a3:zts0:",expectedDigest:"c84ab5a93be130801c54909ccd19f1c71d737315a1550cee0b4a115b6fd687a6"},
  {id:"unicode-keys",value:{"\u{10000}":1,"\ue000":0},expectedEncoding:"o2:s3:\ue000n0000000000000000s4:\u{10000}n3ff0000000000000",expectedDigest:"ad0b977cdfbc27ee6f0972dd99acfdfb326659a0aabb0f592c0eac5bf68b6ce1"},
];
export async function publicationCases(){
  const snapshot=(await buildCoreSnapshot(snapshotFixture())).manifest;
  const candidate={schemaVersion:"ulpin-core-publication/1",state:"candidate",scope:snapshot.context.scope,snapshotDigest:snapshot.inputDigest,geometryDigest:snapshot.geometryDigest,compiler:"qualified-test-compiler/1",assets:[{id:"tile",sha256:"a".repeat(64),bytes:100,mediaType:"model/gltf-binary"}],bindings:[{assetId:"tile",featureId:"building-1",representation:snapshotRef("representation","exterior")}]};
  const cases:{id:string;snapshot:unknown;candidate:unknown;valid:boolean;code?:string}[]=[{id:"candidate-not-activation",snapshot,candidate,valid:true}];
  const bad=(id:string,code:string,change:(c:any)=>void)=>{const copy=JSON.parse(JSON.stringify(candidate));change(copy);cases.push({id,snapshot,candidate:copy,valid:false,code});};
  bad("cannot-self-publish","INVALID_CONTRACT",c=>c.state="active");
  bad("wrong-snapshot-digest","PUBLICATION_SNAPSHOT",c=>c.snapshotDigest="b".repeat(64));
  bad("wrong-geometry-digest","PUBLICATION_SNAPSHOT",c=>c.geometryDigest="b".repeat(64));
  bad("different-scope","ACCESS_SCOPE",c=>c.scope.id="unrelated");
  bad("missing-asset","MISSING_REFERENCE",c=>c.assets=[]);
  bad("stale-feature-revision","STALE_REFERENCE",c=>c.bindings[0].representation.revision=2);
  bad("duplicate-feature","DUPLICATE_REFERENCE",c=>c.bindings.push(c.bindings[0]));
  bad("duplicate-asset","DUPLICATE_REFERENCE",c=>c.assets.push(c.assets[0]));
  return cases;
}
